File: C/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

bool is_valid(long number);
int get_length(long number);
bool check_luhn(long number);
void print_card_type(long number);

int main(void)
{
    long card_number = get_long("Número: ");

    if (is_valid(card_number))
    {
        print_card_type(card_number);
    }
    else
    {
        printf("INVALID\n");
    }
}
bool is_valid(long number)
{
    return check_luhn(number);
}
int get_length(long number)
{
    int length = 0;
    while (number > 0)
    {
        number /= 10;
        length++;
    }
    return length;
}

bool check_luhn(long number)
{
    int sum = 0;
    bool alternate = false;
    while (number > 0)
    {
        int digit = number % 10;
        if (alternate)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }
        sum += digit;
        alternate = !alternate;
        number /= 10;
    }
    return (sum % 10 == 0);
}

void print_card_type(long number)
{
    int length = get_length(number);
    int first_digit = number / (long) pow(10, length - 1);
    int first_two_digits = number / (long) pow(10, length - 2);

    if (length == 15 && (first_two_digits == 34 || first_two_digits == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && (first_two_digits >= 51 && first_two_digits <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && first_digit == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Python/credit.py
from cs50 import get_string


def main():
    card_number = get_string("Number: ")
    if is_valid(card_number):
        card_type = get_card_type(card_number)
        print(card_type)
    else:
        print("INVALID")


def is_valid(card_number):

    sum_doubled_digits = 0
    for i in range(len(card_number) - 2, -1, -2):
        digit = int(card_number[i]) * 2
        sum_doubled_digits += digit // 10 + digit % 10

    sum_other_digits = 0
    for i in range(len(card_number) - 1, -1, -2):
        sum_other_digits += int(card_number[i])

    total_sum = sum_doubled_digits + sum_other_digits
    return total_sum % 10 == 0


def get_card_type(card_number):
    if len(card_number) == 15 and card_number.startswith(("34", "37")):
        return "AMEX"
    elif len(card_number) == 16 and card_number.startswith(("51", "52", "53", "54", "55")):
        return "MASTERCARD"
    elif len(card_number) in (13, 16) and card_number.startswith("4"):
        return "VISA"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()
